nameko_grpc/tracer/dependency.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+import logging
+import sys
+from datetime import datetime
+
+from nameko_tracer import Tracer, constants
+
+from nameko_grpc.constants import Cardinality
+
+
+logger = logging.getLogger(__name__)
+
+GRPC_ADAPTER = {
+    "nameko_grpc.entrypoint.Grpc": ("nameko_grpc.tracer.GrpcEntrypointAdapter")
+}
+
+
+class GrpcTracer(Tracer):
```

Comment:
The difficulty is with the streaming requests and responses. Unlike normal Nameko entrypoints, a streaming method yields multiple times. Similarly for streaming requests, the request object is iterated over yielding multiple items. I think each of these items should result in their own trace.

For a UNARY_STREAM RPC method (i.e. unary request, streaming response) the adapter is called once for the request, once for the "top-level" response (where the result is actually just a generator) and once for every response message in the response stream.

Maybe `nameko-tracer` could support streaming requests/responses as a first-class thing, and then we could drop the subclass here.

Reply:
Matt, you are right, streaming makes it more difficult! Maybe tracer should support it :thinking: :smile: 

Streaming entrypoint yields multiple times but still inside one worker - one entrypoint fire.

If we stick to tracing worker/entrypoint request and response, then the request trace can include just the metadata sent and the response combined information of all responces (and also of all requests in case of stream-unary or stream-stream connection)

I can also see that sometimes "sub-tracing" each chunk/message can be useful. The question is what are the real use-cases and for what scenarios is the streaming used the most? Is it a stream as getting a data flow with getting whats there first before the later stuff is generated? Is it chunking big data sets over the wire? Or is it intended to implement a long lasting communication channel between the client and server which can be achieved in stream-stream scenario? All of these seem to be valid use-cases.

If it is just having a stream or chunking bigger data, I would say having a normal request trace for initiating the call and only one response trace logging results after receiving all responses and after sending all requests would be probably the best fit. The notable work is really done at the end of the entrypoint.

If it is used for some bidirectional messaging implementation, it's true that sub-tracing each may be a better fit.

I quite like having a straight forward tracing of _worker start_ and _worker end_ events, really one entrypoint - one trace. It's not immediately clear that iterating over gRPC requests produces multiple traces, same when yielding multiple times (although this is clearer as there are multiple yield statements present or the yield is in a loop).

Also the pair of caller worker ID and called entrypoint worker ID in the stack trace gets suddenly a bit more complicated as you can get multiple request or response traces per same worker ID. Also these ID pairs are not going to be unique anymore (caller to called worker IDs) Not a big issue, but worth noting.)

When thinking about logging and debugging failures, in a scenario where one sends two successful responses, but fails on generating or sending the third - is it better to have two good traces and one error trace or just logging one trace saying that `entrypoint failed` on third response? Same with streaming a request inside an entrypoint, it's the entrypoint what fails at the end.

Reply:
I just had a look on gRPC [opentracing](https://opentracing.io/)  trying to understand what the best practice might be or what's the standard for tracing gRPC streams.

It seems that a _span_ contains all request and response parts and that one gRPC call results in one span of a trace no matter how many messages its streams had. The finished span is then added to a tracer buffer after the whole remote procedure call was handled.

If `nameko-grpc` supports simple entrypoint/worker level tracing I think it should cover most of the cases .)